docs/openstack/edpm_adoption.md

```diff
@@ -302,6 +267,202 @@ done
+    | jq -e '.[]|select(.type=="Ready" and .status=="True")'
+    oc get novacells --field-selector metadata.name=nova-cell1 -o jsonpath='{.items[0].status.conditions}' \
+    | jq -e '.[]|select(.type=="Ready" and .status=="True")'
+    oc get novaconductors --field-selector metadata.name=nova-cell0-conductor -o jsonpath='{.items[0].status.conditions}' \
+    | jq -e '.[]|select(.type=="Ready" and .status=="True")'
+    oc get novaconductors --field-selector metadata.name=nova-cell1-conductor -o jsonpath='{.items[0].status.conditions}' \
+    | jq -e '.[]|select(.type=="Ready" and .status=="True")'
+    oc get novametadata --field-selector metadata.name=nova-metadata -o jsonpath='{.items[0].status.conditions}' \
+    | jq -e '.[]|select(.type=="Ready" and .status=="True")'
+    oc get novanovncproxies --field-selector metadata.name=nova-cell1-novncproxy -o jsonpath='{.items[0].status.conditions}' \
+    | jq -e '.[]|select(.type=="Ready" and .status=="True")'
+    oc get novaschedulers --field-selector metadata.name=nova-scheduler -o jsonpath='{.items[0].status.conditions}' \
+    | jq -e '.[]|select(.type=="Ready" and .status=="True")'
+    ```
+
+* Remove pre-FFU workarounds for Nova compute EDPM services:
+
+    ```yaml
+    oc apply -f - <<EOF
+    apiVersion: v1
+    kind: ConfigMap
+    metadata:
+      name: nova-compute-ffu
+      namespace: openstack
+    data:
+      20-nova-compute-cell1-ffu-cleanup.conf: |
+        [workarounds]
+        disable_compute_service_check_for_ffu=false
+    EOF
+
+    oc apply -f - <<EOF
+    apiVersion: dataplane.openstack.org/v1beta1
+    kind: OpenStackDataPlaneService
+    metadata:
+      name: nova-compute-ffu
+      namespace: openstack
+    spec:
+      label: nova.compute.ffu
+      configMaps:
+        - nova-compute-ffu
+      playbook: osp.edpm.nova
+    ---
+    apiVersion: dataplane.openstack.org/v1beta1
+    kind: OpenStackDataPlaneDeployment
+    metadata:
+      name: openstack-nova-compute-ffu
+      namespace: openstack
+    spec:
+      nodeSets:
+        - openstack
+      servicesOverride:
+        - nova-compute-ffu
+    EOF
+    ```
+
+* Wait for Nova compute EDPM service to become ready:
+
+    ```bash
+    oc wait --for condition=Ready osdpd/openstack-nova-compute-ffu --timeout=5m
+    ```
+
+* Run Nova DB online migrations to complete FFU:
+
+    ```bash
+    oc exec -it nova-cell0-conductor-0 -- nova-manage db online_data_migrations
+    oc exec -it nova-cell1-conductor-0 -- nova-manage db online_data_migrations
+    ```
+
+* Verify no Nova compute dataplane disruptions during the adoption/upgrade process:
+
+  ```bash
+  $CONTROLLER_SSH sudo podman exec -it libvirt_virtqemud virsh list --all | grep 'instance-00000001   running'
+  ```
+
+* Verify if Nova services control the existing VM instance:
+
+  ```bash
+  openstack server list | grep -qF '| test | ACTIVE |' && openstack server stop test
```

Comment:
something is wrong as at this point nova-compute produces a stack trace
```
2023-11-09 09:55:54.826 2 DEBUG oslo_concurrency.lockutils [None req-b77b398c-9fc4-49fe-95c8-1f6761293777 1d1bd1b129a54c88a4232738e354fbb3 ad151be8d46d451b82f31b39d674565f - - default default] Lock "4191c6c5-7c94-4715-ab88-64b27a7ad2c6" "released" by "nova.compute.manager.ComputeManager.stop_instance.<locals>.do_stop_instance" :: held 3.115s inner /usr/lib/python3.9/site-packages/oslo_concurrency/lockutils.py:423
2023-11-09 09:55:54.978 2 DEBUG oslo_concurrency.lockutils [None req-b77b398c-9fc4-49fe-95c8-1f6761293777 1d1bd1b129a54c88a4232738e354fbb3 ad151be8d46d451b82f31b39d674565f - - default default] Acquiring lock "compute_resources" by "nova.compute.resource_tracker.ResourceTracker.update_usage" inner /usr/lib/python3.9/site-packages/oslo_concurrency/lockutils.py:404
2023-11-09 09:55:54.980 2 DEBUG oslo_concurrency.lockutils [None req-b77b398c-9fc4-49fe-95c8-1f6761293777 1d1bd1b129a54c88a4232738e354fbb3 ad151be8d46d451b82f31b39d674565f - - default default] Lock "compute_resources" acquired by "nova.compute.resource_tracker.ResourceTracker.update_usage" :: waited 0.002s inner /usr/lib/python3.9/site-packages/oslo_concurrency/lockutils.py:409
2023-11-09 09:55:55.070 2 DEBUG nova.compute.provider_tree [None req-b77b398c-9fc4-49fe-95c8-1f6761293777 1d1bd1b129a54c88a4232738e354fbb3 ad151be8d46d451b82f31b39d674565f - - default default] Inventory has not changed in ProviderTree for provider: d3d2be51-a0b9-4538-a298-62280a52fece update_inventory /usr/lib/python3.9/site-packages/nova/compute/provider_tree.py:180
2023-11-09 09:55:55.098 2 DEBUG nova.scheduler.client.report [None req-b77b398c-9fc4-49fe-95c8-1f6761293777 1d1bd1b129a54c88a4232738e354fbb3 ad151be8d46d451b82f31b39d674565f - - default default] Inventory has not changed for provider d3d2be51-a0b9-4538-a298-62280a52fece based on inventory data: {'VCPU': {'total': 8, 'reserved': 0, 'min_unit': 1, 'max_unit': 8, 'step_size': 1, 'allocation_ratio': 16.0}, 'MEMORY_MB': {'total': 19744, 'reserved': 512, 'min_unit': 1, 'max_unit': 19744, 'step_size': 1, 'allocation_ratio': 1.0}, 'DISK_GB': {'total': 69, 'reserved': 1, 'min_unit': 1, 'max_unit': 69, 'step_size': 1, 'allocation_ratio': 1.0}} set_inventory_for_provider /usr/lib/python3.9/site-packages/nova/scheduler/client/report.py:940
2023-11-09 09:55:55.103 2 DEBUG oslo_concurrency.lockutils [None req-b77b398c-9fc4-49fe-95c8-1f6761293777 1d1bd1b129a54c88a4232738e354fbb3 ad151be8d46d451b82f31b39d674565f - - default default] Lock "compute_resources" "released" by "nova.compute.resource_tracker.ResourceTracker.update_usage" :: held 0.123s inner /usr/lib/python3.9/site-packages/oslo_concurrency/lockutils.py:423
2023-11-09 09:55:55.104 2 INFO nova.compute.manager [None req-b77b398c-9fc4-49fe-95c8-1f6761293777 1d1bd1b129a54c88a4232738e354fbb3 ad151be8d46d451b82f31b39d674565f - - default default] [instance: 4191c6c5-7c94-4715-ab88-64b27a7ad2c6] Successfully reverted task state from powering-off on failure for instance.
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server [None req-b77b398c-9fc4-49fe-95c8-1f6761293777 1d1bd1b129a54c88a4232738e354fbb3 ad151be8d46d451b82f31b39d674565f - - default default] Exception during message handling: nova.exception.InstanceNotFound: Instance 4191c6c5-7c94-4715-ab88-64b27a7ad2c6 could not be found.
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server Traceback (most recent call last):
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server   File "/usr/lib/python3.9/site-packages/nova/virt/libvirt/host.py", line 690, in _get_domain
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server     return conn.lookupByUUIDString(instance.uuid)
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server   File "/usr/lib/python3.9/site-packages/eventlet/tpool.py", line 193, in doit
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server     result = proxy_call(self._autowrap, f, *args, **kwargs)
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server   File "/usr/lib/python3.9/site-packages/eventlet/tpool.py", line 151, in proxy_call
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server     rv = execute(f, *args, **kwargs)
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server   File "/usr/lib/python3.9/site-packages/eventlet/tpool.py", line 132, in execute
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server     six.reraise(c, e, tb)
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server   File "/usr/lib/python3.9/site-packages/six.py", line 709, in reraise
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server     raise value
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server   File "/usr/lib/python3.9/site-packages/eventlet/tpool.py", line 86, in tworker
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server     rv = meth(*args, **kwargs)
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server   File "/usr/lib64/python3.9/site-packages/libvirt.py", line 5008, in lookupByUUIDString
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server     raise libvirtError('virDomainLookupByUUIDString() failed')
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server libvirt.libvirtError: Domain not found: no domain with matching uuid '4191c6c5-7c94-4715-ab88-64b27a7ad2c6'
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server 
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server During handling of the above exception, another exception occurred:
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server 
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server Traceback (most recent call last):
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server   File "/usr/lib/python3.9/site-packages/oslo_messaging/rpc/server.py", line 165, in _process_incoming
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server     res = self.dispatcher.dispatch(message)
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server   File "/usr/lib/python3.9/site-packages/oslo_messaging/rpc/dispatcher.py", line 309, in dispatch
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server     return self._do_dispatch(endpoint, method, ctxt, args)
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server   File "/usr/lib/python3.9/site-packages/oslo_messaging/rpc/dispatcher.py", line 229, in _do_dispatch
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server     result = func(ctxt, **new_args)
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server   File "/usr/lib/python3.9/site-packages/nova/exception_wrapper.py", line 71, in wrapped
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server     _emit_versioned_exception_notification(
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server   File "/usr/lib/python3.9/site-packages/oslo_utils/excutils.py", line 227, in __exit__
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server     self.force_reraise()
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server   File "/usr/lib/python3.9/site-packages/oslo_utils/excutils.py", line 200, in force_reraise
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server     raise self.value
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server   File "/usr/lib/python3.9/site-packages/nova/exception_wrapper.py", line 63, in wrapped
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server     return f(self, context, *args, **kw)
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server   File "/usr/lib/python3.9/site-packages/nova/compute/manager.py", line 186, in decorated_function
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server     LOG.warning("Failed to revert task state for instance. "
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server   File "/usr/lib/python3.9/site-packages/oslo_utils/excutils.py", line 227, in __exit__
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server     self.force_reraise()
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server   File "/usr/lib/python3.9/site-packages/oslo_utils/excutils.py", line 200, in force_reraise
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server     raise self.value
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server   File "/usr/lib/python3.9/site-packages/nova/compute/manager.py", line 157, in decorated_function
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server     return function(self, context, *args, **kwargs)
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server   File "/usr/lib/python3.9/site-packages/nova/compute/utils.py", line 1439, in decorated_function
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server     return function(self, context, *args, **kwargs)
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server   File "/usr/lib/python3.9/site-packages/nova/compute/manager.py", line 203, in decorated_function
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server     return function(self, context, *args, **kwargs)
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server   File "/usr/lib/python3.9/site-packages/nova/compute/manager.py", line 3381, in stop_instance
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server     do_stop_instance()
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server   File "/usr/lib/python3.9/site-packages/oslo_concurrency/lockutils.py", line 414, in inner
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server     return f(*args, **kwargs)
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server   File "/usr/lib/python3.9/site-packages/nova/compute/manager.py", line 3369, in do_stop_instance
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server     self._power_off_instance(instance, clean_shutdown)
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server   File "/usr/lib/python3.9/site-packages/nova/compute/manager.py", line 3076, in _power_off_instance
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server     self.driver.power_off(instance, timeout, retry_interval)
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server   File "/usr/lib/python3.9/site-packages/nova/virt/libvirt/driver.py", line 4099, in power_off
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server     self._clean_shutdown(instance, timeout, retry_interval)
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server   File "/usr/lib/python3.9/site-packages/nova/virt/libvirt/driver.py", line 4059, in _clean_shutdown
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server     guest = self._host.get_guest(instance)
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server   File "/usr/lib/python3.9/site-packages/nova/virt/libvirt/host.py", line 674, in get_guest
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server     return libvirt_guest.Guest(self._get_domain(instance))
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server   File "/usr/lib/python3.9/site-packages/nova/virt/libvirt/host.py", line 694, in _get_domain
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server     raise exception.InstanceNotFound(instance_id=instance.uuid)
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server nova.exception.InstanceNotFound: Instance 4191c6c5-7c94-4715-ab88-64b27a7ad2c6 could not be found.
2023-11-09 09:55:55.107 2 ERROR oslo_messaging.rpc.server 
2023-11-09 09:55:56.053 2 DEBUG ovsdbapp.backend.ovs_idl.vlog [-] [POLLIN] on fd 19 __log_wakeup /usr/lib64/python3.9/site-packages/ovs/poller.py:263
```
Then I cannot start the instance up again
```
[gibi@osp-dev-01 ~]$ openstack server start test
Cannot 'start' instance 4191c6c5-7c94-4715-ab88-64b27a7ad2c6 while it is in vm_state active (HTTP 409) (Request-ID: req-eb72c115-a9da-4e47-acbc-2457ddbf7607)
command terminated with exit code 1
```

Reply:
this is weird, I haven't observed that during my testing :)